DOC Document new session save handlers (#783)
silverstripe/developer-docs: en/02_Developer_Guides/18_Cookies_And_Sessions/02_Sessions.md

  save_handler: null
```

You can also set it to the FQCN or injector service name of any session handler that implements [`SessionHandlerInterface`](https://www.php.net/manual/en/class.sessionhandlerinterface.php). `silverstripe/framework` comes with several non-blocking session save handlers you can use.

> [!WARNING]

> > Note that in edge case scenarios, for example if your application wants to modify a session value *based on the value that is already set* and must do so for each request, non-blocking sessions may cause unexpected results.

Reviewer: ```suggestion
> In edge case scenarios, for example if your application wants to modify a session value *based on the value that is already set* and must do so for each request, non-blocking sessions may cause unexpected results.
```

Reviewer 2: Done

#### `FileSessionHandler`

The default file-based session handler for PHP holds a lock on the session file while the session is open. This means that multiple concurrent requests from the same user have to wait for one another to finish processing after a session has been started. This includes AJAX requests.

To resolve this problem, Silverstripe CMS comes with [`FileSessionHandler`](api:SilverStripe\Control\SessionHandler\FileSessionHandler).

`FileSessionHandler` differs from the default PHP file session handler in the following ways:

1. It doesn't lock the session file, and therefore doesn't block concurrent requests.
1. The [`Session.timeout`](api:SilverStripe\Control\Session->timeout) configuration property is used as the source of truth for the lifetime of session files (see [session lifetime](#session-lifetime) above).
1. If there are problems reading or writing to session files, the [default logging service](/developer_guides/debugging/error_handling/) is used to log them.

#### `CacheSessionHandler`

> If you want a more performant session save handler, you can use the [`CacheSessionHandler`](api:SilverStripe\Control\SessionHandler\CacheSessionHandler). This session save handler can use any cache adapter that implements `Psr\SimpleCache\CacheInterface`, though we recommend specifically using an in-memory cache adapter that gets instantiated from a factory implementing [`InMemoryCacheFactory`](api:SilverStripe\Core\Cache\InMemoryCacheFactory).

Reviewer: ```suggestion
If you want a more performant session save handler, you can use the [`CacheSessionHandler`](api:SilverStripe\Control\SessionHandler\CacheSessionHandler). This session save handler can use any cache that implements the PSR-16 [`Psr\SimpleCache\CacheInterface`](https://www.php-fig.org/psr/psr-16/#21-cacheinterface), though we recommend specifically using an in-memory cache that gets instantiated from a factory implementing [`InMemoryCacheFactory`](api:SilverStripe\Core\Cache\InMemoryCacheFactory), such as [`MemcachedCacheFactory`](api:SilverStripe\Core\Cache\MemcachedCacheFactory) or [`RedisCacheFactory`](api:SilverStripe\Core\Cache\RedisCacheFactory).
```

Reviewer 2: I'll keep "in-memory cache adapter" because in this case it's referring specifically to adapters as documented in https://docs.silverstripe.org/en/6/developer_guides/performance/cache_adapters/#adding-an-in-memory-cache-adapter

Reviewer 2: Done but removed the bit further down which served the same purpose of pointing out redis and memcached. It was

```md
For example, if you have Redis you can use [`RedisCacheFactory`](api:SilverStripe\Core\Cache\RedisCacheFactory), and if you have Memcached you can use [`MemcachedCacheFactory`](SilverStripe\Core\Cache\MemcachedCacheFactory).
```

> [!NOTE]

> > Although this is using a cache adapter, sessions won't be cleared when flushing the site.

Reviewer: ```suggestion
> [!NOTE]
> When using a PSR-16 compatible cache, sessions won't be cleared when flushing the site.
```

Reviewer 2: It doesn't matter whether they're PSR-16 or not. I'll remove the word "adapter" which is what I assume you're really after.

> You can define the factory which is used to instantiate the cache handler by setting the `SS_SESSION_CACHE_FACTORY` environment variable, or by setting the following YAML configuration:

Reviewer: ```suggestion
You can define the factory which is used to instantiate the cache by setting the `SS_SESSION_CACHE_FACTORY` environment variable, or by setting the following YAML configuration:
```

Reviewer 2: Done

```yml
---
After: '#session-handlers'
---
SilverStripe\Core\Injector\Injector:
  Psr\SimpleCache\CacheInterface.session-handler:
    factory: 'App\Session\MyCacheFactory'
```

> For example, if you have Redis you can use [`RedisCacheFactory`](api:SilverStripe\Core\Cache\RedisCacheFactory), and if you have Memcached you can use [`MemcachedCacheFactory`](SilverStripe\Core\Cache\MemcachedCacheFactory).

Reviewer: ```suggestion
For example, if you have Redis you can use `RedisCacheFactory`, and if you have Memcached you can use `MemcachedCacheFactory`.
```
API docs are linked above in my recommended changes

Reviewer 2: Removed as redundant to https://github.com/silverstripe/developer-docs/pull/783#discussion_r2224128967

See [cache adapters](/developer_guides/performance/cache_adapters/) for any additional details required to use those cache factories.

#### `DatabaseSessionHandler`

The [`DatabaseSessionHandler`](api:SilverStripe\Control\SessionHandler\DatabaseSessionHandler) class lets you store session data in the database.

This provides a low barrier to sharing your sessions across multiple servers, e.g. in a horizontally-scaled hosting scenario.

> You can change the name of the table used by setting [`DatabaseSessionHandler.table_name`](api:SilverStripe\Control\SessionHandler\DatabaseSessionHandler->table_name) to the new table name. Note that changing the table name after sessions have already been stored in the old table will result in those sessions being invalidated, unless you manually migrate them to the new table.

Reviewer: ```suggestion
You can change the name of the table used by setting [`DatabaseSessionHandler.table_name`](api:SilverStripe\Control\SessionHandler\DatabaseSessionHandler->table_name) from the default of `_sessions` to a new table name.

> [!IMPORTANT]
> Changing the table name after sessions have already been stored in the old table will result in those sessions being invalidated, unless you manually migrate them to the new table
```

Reviewer 2: We don't tend so say "from the default of `<default value>`" in docs elsewhere from what I can recall so I'll not mention `_sessions` explicitly. But will make the other changes.

Will use `WARNING` instead of `IMPORTANT` as per https://docs.silverstripe.org/en/6/contributing/documentation/#callout-blocks

Reviewer 2: Done
